internal/charmstore/migrations_test.go

```diff
@@ -793,6 +831,692 @@ func (s *migrationsSuite) TestPopulateWriteACLSomeUpdates(c *gc.C) {
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"dalek"`),
+		},
+	})
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id5,
+		BaseURL:             baseURL(id5),
+		BlobHash:            "rails-1",
+		User:                id5.User,
+		Name:                id5.Name,
+		Revision:            id5.Revision,
+		Series:              id5.Series,
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"ace"`),
+		},
+	})
+	s.checkEntity(c, &mongodoc.Entity{
+		URL:                 id6,
+		BaseURL:             baseURL(id6),
+		BlobHash:            "django-2",
+		User:                id6.User,
+		Name:                id6.Name,
+		Revision:            id6.Revision,
+		Series:              id6.Series,
+		PromulgatedRevision: -1,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+	})
+
+	s.checkCount(c, s.db.BaseEntities(), 4)
+	s.checkBaseEntity(c, &mongodoc.BaseEntity{
+		URL:         baseURL(id2),
+		User:        id2.User,
+		Name:        id2.Name,
+		Public:      true,
+		ACLs:        mongodoc.ACL{},
+		Promulgated: 0,
+	})
+	s.checkBaseEntity(c, &mongodoc.BaseEntity{
+		URL:         baseURL(id3),
+		User:        id3.User,
+		Name:        id3.Name,
+		Public:      true,
+		ACLs:        mongodoc.ACL{},
+		Promulgated: 1,
+	})
+	s.checkBaseEntity(c, &mongodoc.BaseEntity{
+		URL:         baseURL(id4),
+		User:        id4.User,
+		Name:        id4.Name,
+		Public:      true,
+		ACLs:        mongodoc.ACL{},
+		Promulgated: 0,
+	})
+	s.checkBaseEntity(c, &mongodoc.BaseEntity{
+		URL:         baseURL(id5),
+		User:        id5.User,
+		Name:        id5.Name,
+		Public:      true,
+		ACLs:        mongodoc.ACL{},
+		Promulgated: 0,
+	})
+}
+
+func (s *migrationsSuite) TestPopulatePromulgatedEntitiesSomeUpdates(c *gc.C) {
+	s.patchMigrations(c, getMigrations("populate promulgated entities"))
+	id1 := charm.MustParseReference("trusty/django-42")
+	id2 := charm.MustParseReference("~who/utopic/rails-47")
+	id3 := charm.MustParseReference("~who/trusty/django-42")
+	id4 := charm.MustParseReference("~dalek/trusty/django-42")
+	id5 := charm.MustParseReference("~ace/utopic/rails-47")
+	id6 := charm.MustParseReference("~who/trusty/django-41")
+	id7 := charm.MustParseReference("trusty/django-41")
+	err := s.db.Entities().Insert(&entity2{
+		URL:      id1,
+		BaseURL:  baseURL(id1),
+		BlobHash: "django-1",
+		Name:     id1.Name,
+		User:     id1.User,
+		Series:   id1.Series,
+		Revision: id1.Revision,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+	})
+	c.Assert(err, gc.IsNil)
+	err = s.db.Entities().Insert(&entity2{
+		URL:      id2,
+		BaseURL:  baseURL(id2),
+		BlobHash: "rails-1",
+		Name:     id2.Name,
+		User:     id2.User,
+		Series:   id2.Series,
+		Revision: id2.Revision,
+		ExtraInfo: map[string][]byte{
+			"bzr-owner": []byte(`"who"`),
+		},
+	})
+	c.Assert(err, gc.IsNil)
+	err = s.db.Entities().Insert(&entity2{
```

Comment:
Perhaps, there are functions for doing similar things already. As we get more migrations we will probably need a more general approach.